ambry-store/src/main/java/com/github/ambry/store/StorageManager.java

```diff
@@ -578,29 +578,43 @@ private class PartitionStateChangeListenerImpl implements PartitionStateChangeLi
-              "New replica " + partitionName + " is not found in clustermap for " + currentNode, ReplicaNotFound);
-        }
-        // Attempt to add store into storage manager. If store already exists on disk (but not in clustermap), make
-        // sure old store of this replica is deleted (this store may be created in previous replica addition but failed
-        // at some point). Then a brand new store associated with this replica should be created and started.
-        if (!addBlobStore(replicaToAdd)) {
-          // We have decreased the available disk space in HelixClusterManager#getDiskForBootstrapReplica. Increase it
-          // back since addition of store failed.
-          replicaToAdd.getDiskId().increaseAvailableSpaceInBytes(replicaToAdd.getCapacityInBytes());
-          logger.error("Failed to add store {} into storage manager", partitionName);
-          throw new StateTransitionException("Failed to add store " + partitionName + " into storage manager",
-              ReplicaOperationFailure);
-        }
+
+        ReplicaId replicaToAdd;
+        boolean replicaAdded = false;
+        do {
+          // there can be two scenarios:
+          // 1. this is the first time to add new replica onto current node;
+          // 2. last replica addition failed at some point before updating InstanceConfig in Helix
+          // In either case, we should add replica to current node by calling "addBlobStore(ReplicaId replica)"
+          replicaToAdd = clusterMap.getBootstrapReplica(partitionName, currentNode);
+          if (replicaToAdd == null) {
+            logger.error("No new replica found for partition {} in cluster map", partitionName);
+            throw new StateTransitionException(
+                "New replica " + partitionName + " is not found in clustermap for " + currentNode, ReplicaNotFound);
+          }
+          // Attempt to add store into storage manager. If store already exists on disk (but not in clustermap), make
+          // sure old store of this replica is deleted (this store may be created in previous replica addition but failed
+          // at some point). Then a brand new store associated with this replica should be created and started.
+          if (!addBlobStore(replicaToAdd)) {
+            // We have decreased the available disk space in HelixClusterManager#getDiskForBootstrapReplica. Increase it
+            // back since addition of store failed.
+            replicaToAdd.getDiskId().increaseAvailableSpaceInBytes(replicaToAdd.getCapacityInBytes());
+            if (!clusterMap.isDataNodeInFullAutoMode(currentNode)) {
+              logger.error("Failed to add store {} into storage manager", partitionName);
+              throw new StateTransitionException("Failed to add store " + partitionName + " into storage manager",
+                  ReplicaOperationFailure);
+            } else {
+              // TODO: Delete any files added in store and reserve directory
```

Comment:
we should do this, otherwise, it would just take some disk space and won't let it go before we restart. I suppose we can just add a new method in disk manager to clean up the "unexpected dirs". I have a PR to do that, let's merge these two PRs and then use the method in my PR to remove the partition directories later.

Reply:
Sure. I will merge this PR and put up a new PR